### ISSUE TITLE
STXXL corrupts name lists when they are large

### ISSUE THREAD

**Issue by danpat** [Project-OSRM/osrm-backend#1808]
TL;DR: `STXXLVector<std::string>` is producing undefined behaviour when things get paged out.  The STXXL container types only support POD objects.  `extraction_containers.hpp` has an `STXXLVector<string>` that breaks when large numbers of strings are inserted that trigger paging behaviour.  It works if not enough data is inserted to trigger paging, which seems to be what we've been doing up to now.

If you use the `name` attribute on ways as a storage mechanism for other metadata, this generates a much larger `.names` file than if just regular street names are used.

Under circumstances where there are very large numbers of unique names (in our test data set, around 67M unique names), STXXL seems to corrupt the name strings upon retrieval, which leads to a segfault at the end of the first stage of `osrm-extract`.

The stack trace looks like this:

```
[extractor] Sorting used ways         ... ok, after 35.3939s
[extractor] Sorting 0 restriction. by from... ok, after 0.000661s
[extractor] Fixing restriction starts ... ok, after 0s
[extractor] Sorting restrictions. by to  ... ok, after 0.000308s
[extractor] Fixing restriction ends   ... ok, after 0s
[info] usable restrictions: 0
[extractor] writing street name index ...
Program received signal SIGSEGV, Segmentation fault.
(gdb) bt
   #0  __memcpy_sse2_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S:36
   #1  0x00007ffff607d430 in std::basic_streambuf<char, std::char_traits<char> >::xsputn(char const*, long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
   #2  0x00007ffff6074772 in std::ostream::write(char const*, long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
   #3  0x00000000004df7a3 in ExtractionContainers::WriteNames(std::string const&) const ()
   #4  0x000000000054a4a9 in ExtractionContainers::PrepareData(std::string const&, std::string const&, std::string const&, lua_State*) ()
   #5  0x000000000049dcc7 in extractor::run() ()
   #6  0x000000000046f965 in main ()
   (gdb)
```

After lots of debugging, it turns out that when we retrieve the `name` values from the `external_name_list` here:

https://github.com/Project-OSRM/osrm-backend/blob/c38183423f82bd1e0568af04201f12599966d2fe/extractor/extraction_containers.cpp#L136-L138

adding some debugging to this loop:

    std::cout << "Writing string \"" << temp_string.c_str() << "\" length: " << temp_string.length() << ", " << string_length << std::endl;

returns values like this:

    Writing string "35090902" length: 140385527561280, 255
    Writing string "35162624" length: 140381485532832, 255
    Writing string "35059731" length: 140383840544112, 255
    Writing string "U�"��" length: 0, 0
    Writing string "35276273" length: 140385414743664, 255
    Writing string "35169595" length: 140384273544512, 255
    Writing string "35222782" length: 140511914369232, 255
    Writing string "35122530" length: 140383979460432, 255

Where those `length` values are clearly incorrect.  I confirmed that the strings being *added* to the `STXXLVector<string>` were correct here:

https://github.com/Project-OSRM/osrm-backend/blob/c38183423f82bd1e0568af04201f12599966d2fe/extractor/extractor_callbacks.cpp#L159

The problem appears to be the use of `std::string` along with `STXXLVector`.  The STXXL FAQ states:

> STXXL container types like stxxl::vector can be parameterized only with a value type that is a POD (i. e. no virtual functions, no user-defined copy assignment/destructor, etc.) and does not contain references (including pointers) to internal memory. Usually, "complex" data types do not satisfy this requirements.

There are plenty of comments online backing this up, specifically that `std::string` will probably not work:

http://sourceforge.net/p/stxxl/discussion/446474/thread/7b904a22/

It seems likely that we've managed to avoid writing anything out to disk until now, so the `std::string` values have remained in-memory undamaged.  However, by massively increasing the number of unique `name` values, we've exposed this undefined behaviour.

The "simple" fix is to replace the `STXXL<string> external_name_list` in `extraction_containers.hpp` with a standard `std::vector<string>`.  It seems that we haven't actually been using STXXL's disk-paging here anyway (because it would've broken).

https://github.com/Project-OSRM/osrm-backend/blob/c38183423f82bd1e0568af04201f12599966d2fe/extractor/extraction_containers.hpp#L67

Note, we're *already* storing every name in an in-memory `std::unordered_map` here:

https://github.com/Project-OSRM/osrm-backend/blob/c38183423f82bd1e0568af04201f12599966d2fe/extractor/extractor_callbacks.hpp#L59

**Comment by rparanjpe-tesla** [comment]
*It seems likely that we've managed to avoid writing anything out to disk until now, so the std::string values have remained in-memory undamaged.*

I'm not sure about this because when we ran osrm-extract with our .stxxl config file to use memory only, we got the same errors.
```
disk=/path/to/stxxl,120000,memory
```

**Comment by danpat** [comment]
The problem is likely being caused by when STXXL serializes/deserializes `std::string` objects when it decides that it needs to page some objects out.  Paging out to disk or to the `memory` backend has the same effect here: serializing/deserializing a `std::string` breaks it.

**Comment by daniel-j-h** [comment]
I would suggest upstreaming a few `static_assert(is_pod_v<T>, "");` for sanity.

**Comment by danpat** [comment]
Closed in #1809
